dds/DCPS/transport/rtps_udp/RtpsUdpDataLink.cpp

```diff
@@ -1694,60 +1683,57 @@ RtpsUdpDataLink::received(const RTPS::HeartBeatSubmessage& heartbeat,
       RtpsReaderMap::const_iterator riter = readers_.find(readerid);
       if (riter == readers_.end()) {
         // Reader has no associations.
-        interesting_ack_nacks_.insert(InterestingAckNack(writerid, readerid, pos->second.address));
+        send_interesting_ack_nack(writerid, readerid, heartbeat.count.value, meta_submessages);
       } else if (!riter->second->has_writer(writerid)) {
         // Reader is not associated with this writer.
-        interesting_ack_nacks_.insert(InterestingAckNack(writerid, readerid, pos->second.address));
+        send_interesting_ack_nack(writerid, readerid, heartbeat.count.value, meta_submessages);
       }
       pos->second.last_activity = now;
       if (pos->second.status == InterestingRemote::DOES_NOT_EXIST) {
         callbacks.push_back(pos->second);
         pos->second.status = InterestingRemote::EXISTS;
       }
     }
-
-    schedule_acknack = !interesting_ack_nacks_.empty();
   }
+  queue_or_send_submessages(meta_submessages);
 
   for (size_t i = 0; i < callbacks.size(); ++i) {
     callbacks[i].listener->writer_exists(src, callbacks[i].localid);
   }
 
-  if (schedule_acknack) {
-    heartbeat_reply_.enable(normal_heartbeat_response_delay_);
-  }
-
   datareader_dispatch(heartbeat, src_prefix, directed, &RtpsReader::process_heartbeat_i);
 }
 
-bool
+void
 RtpsUdpDataLink::RtpsReader::process_heartbeat_i(const RTPS::HeartBeatSubmessage& heartbeat,
                                                  const RepoId& src,
                                                  bool directed,
                                                  MetaSubmessageVec& meta_submessages)
 {
-  ACE_GUARD_RETURN(ACE_Thread_Mutex, g, mutex_, false);
+  ACE_GUARD(ACE_Thread_Mutex, g, mutex_);
 
   RtpsUdpDataLink_rch link = link_.lock();
 
   if (!link) {
-    return false;
+    return;
   }
 
   GuardType guard(link->strategy_lock_);
   if (link->receive_strategy() == 0) {
-    return false;
+    return;
   }
 
   const WriterInfoMap::iterator wi = remote_writers_.find(src);
   if (wi == remote_writers_.end()) {
-    return false;
+    return;
   }
 
   const WriterInfo_rch& writer = wi->second;
 
   if (!compare_and_update_counts(heartbeat.count.value, writer->heartbeat_recvd_count_)) {
-    return false;
+    VDBG((LM_WARNING, "(%P|%t) RtpsUdpDataLink::process_heartbeat_i "
```

Comment:
Does VDBG buy us anything here? We weren't historically using it in RtpsUdpDatalink.

We really need to get a better logging setup for RTPS/UDP... ability to filter by message type & builtin vs user entity at the very least...

Reply:
I did it to be consistent with similar code.